server/ServerApp.cpp

```diff
@@ -875,32 +859,96 @@ void ServerApp::LoadSPGameInit(const std::vector<PlayerSaveGameData>& player_sav
     LoadGameInit(player_save_game_data, player_id_to_save_game_data_index, server_save_game_data);
 }
 
+namespace {
+        /** Check that \p path is a file or directory in the server save
+        directory. */
+    bool IsInServerSaveDir(const fs::path& path) {
+
+        fs::path server_dir = GetServerSaveDir();
+        if (!fs::exists(server_dir) || !fs::is_directory(server_dir))
+            return false;
+
+        if (!fs::exists(path))
+            return false;
+
+        // Resolve any symbolic links, dots or dot-dots
+        auto canon_server = fs::canonical(server_dir);
+        auto canon_path = fs::canonical(path);
+
+        // Paths shorter than the server save dir are not in the server dir
+        auto save_dir_length = std::distance(canon_server.begin(), canon_server.end());
+        auto path_length = std::distance(canon_path.begin(), canon_path.end());
+        if (path_length < save_dir_length)
+            return false;
+
+        // Check that the whole server save dir path matches the test path
+        return std::equal(canon_server.begin(), canon_server.end(), path.begin());
+    }
+
+    /// Generates information on the subdirectories of \p directory
+    std::vector<std::string> ListSaveSubdirectories(const fs::path& directory) {
+        std::vector<std::string> list;
+        if (!fs::is_directory(directory))
+            return list;
+
+        std::string server_dir_str = fs::canonical(GetServerSaveDir()).generic_string();
```

Comment:
Could probably call `PathString` here, and `FilenameToPath` whenever taking a string received in a message from the client that needs to be converted back into a path.

Reply:
Thanks for looking into that issue.

I've made the changes.  It might now also work on Windows